README.md
# 1337 Translator

1337 is an alternative alphabet for the English language that uses various combinations of characters to replace letters. The translation alphabet is located at ```src/alphabet.json``` and is based off of [Wikipedia's 1337 translation table](http://en.wikipedia.org/wiki/Leet#Orthography).

## Installation

No automated installation process has been implemented as of yet. It is planned for the future.

## Usage

1337 Translator comes with an array of command line options to customize the translation process.

### Input Options

Only one of the following input options can be selected at once. If no input option is selected, you will be prompted to enter the input by keyboard.

    -i INPUT
    --input INPUT

Enter the string to be used directly into the command line (replace INPUT).

    -c
    --clipboard

Take the input directly from the clipboard.

    -f FILE
    --file FILE

Take the input directly from the specified file (replace FILE).

### Fakeout Options

Sometimes a direct translation isn't enough; if you want to obfuscate the 1337 even more, try one or more of the following. You will be able to see the sequence of changes that the preprocessor makes as it applies these options, and the output is colored to assist in viewing.

    -l INTERVAL
    --fake-letters INTERVAL

Insert a random phoney letter at each interval. For example, ```my hovercraft is full of eels``` with the option ```-l4``` or ```--fake-letters 4``` would output something to the effect of ```my hoevercnraft yis fupll of peels```.

    -w INTERVAL
    --fake-words INTERVAL

Insert a random phoney word at each interval, in a similar fashion to ```--fake-letters```. The list of words selected from is at ```src/fakewords.txt```, modify it to your liking.

    -s
    --full-stop

Translate all full stops to STOP, as in a telegram. ```First sentence. Second sentence.``` would give ```First sentence STOP Second sentence STOP```. This may also help greatly to differentiate periods from other 1337 symbols when it is being decoded. The STOPs do not count toward the interval for ```--fake-words```.

### Output Options

These options allow output to places other than the console output. As many of these may be selected as you please.

    -C
    --out-clipboard

Output the translation to the clipboard.

    -o FILE
    --output FILE

Output the translation to the designated file (replace FILE).

### Miscellaneous Options

    -h
    --help

Shows a set of instructions for command line usage of 1337 Translator.

    --version

Output the version of 1337 Translator that you are using and then exit.

    -q
    --quiet
    --silent

Does not display anything on the console output during the translation. It is recommended to use an output option with this, otherwise you won't be able to see the translation! This will not silence any error messages, should they arise.

    -t
    --no-trim

By default, whitespace is trimmed from the beginning and end of the input; this includes spaces, tabs, newlines, etc. Use this option if you wish to not disturb the whitespace.

## Contributing

Don't be afraid to send a pull request if you have a contribution to make to the code. Even if you are unable to contribute to the code of 1337 Translator, reports of any bugs that you may find are appreciated.


CHANGELOG.md
# 1337 Translator Change Log

All changes to 1337 Translator will be logged in this file with each release.

## 0.1

This is the initial release of 1337 Translator.

### Goals for Next Release

 + Correct any bugs discovered and get user feedback from this release.

## 0.2

**Currently in the trunk.**

### Bug Fixes

 + Fixed only one output option working at a time

### Tweaks

 + ```--full-stop``` will no longer display a result output if there were no full stops to process.

### New Features

 + Text input can now be entered at the command line.
 + The preprocessing display now has colors to aid in viewing the changes.
 + There is now a version option to view the current version of this program.


src/pyperclip.py
# Pyperclip v1.3
# A cross-platform clipboard module for Python.
# By Al Sweigart <al@coffeeghost.net>
# Modified for 1337 Translator by Vinny Diehl <gbchaosmaster926@gmail.com>

# Usage:
#   import pyperclip
#   pyperclip.copy('The text to be copied to the clipboard.')
#   spam = pyperclip.paste()

# On Mac, this module makes use of the pbcopy and pbpaste commands,
# which should come with the os.
# On Linux, this module makes use of the xclip command, which should
# come with the os. Otherwise run "sudo apt-get install xclip"

# Copyright (c) 2010, Albert Sweigart
# All rights reserved.

# BSD-style license:

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#     * Neither the name of the pyperclip nor the
#       names of its contributors may be used to endorse or promote products
#       derived from this software without specific prior written permission.

# THIS SOFTWARE IS PROVIDED BY Albert Sweigart "AS IS" AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL Albert Sweigart BE LIABLE FOR ANY
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# Change Log:
# 1.2 Use the platform module to help determine OS.
# 1.3 Changed ctypes.windll.user32.OpenClipboard(None) to
#     ctypes.windll.user32.OpenClipboard(0), after some people ran
#     into some TypeError

# 2012-05-31 gbchaosmaster - Making this support UTF-8. Adds Django dependency.

import platform, os
from django.utils.encoding import smart_str

def winGetClipboard():
    ctypes.windll.user32.OpenClipboard(0)
    pcontents = ctypes.windll.user32.GetClipboardData(1) # 1 is CF_TEXT
    data = ctypes.c_char_p(pcontents).value
    ctypes.windll.user32.CloseClipboard()
    return data

def winSetClipboard(text):
    GMEM_DDESHARE = 0x2000
    ctypes.windll.user32.OpenClipboard(0)
    ctypes.windll.user32.EmptyClipboard()
    try:
        # Works on Python 2 (bytes() only takes one argument)
        hCd = ctypes.windll.kernel32.GlobalAlloc(GMEM_DDESHARE,
                                                 len(bytes(smart_str(text)))+1)
    except TypeError:
        # Works on Python 3 (bytes() requires an encoding)
        hCd = ctypes.windll.kernel32.GlobalAlloc(GMEM_DDESHARE,
                                                 len(bytes(smart_str(text),
                                                           'ascii')) + 1)

    pchData = ctypes.windll.kernel32.GlobalLock(hCd)
    try:
        # Works on Python 2 (bytes() only takes one argument)
        ctypes.cdll.msvcrt.strcpy(ctypes.c_char_p(pchData),
                                  bytes(smart_str(text)))
    except TypeError:
        # Works on Python 3 (bytes() requires an encoding)
        ctypes.cdll.msvcrt.strcpy(ctypes.c_char_p(pchData),
                                  bytes(smart_str(text), 'ascii'))
    ctypes.windll.kernel32.GlobalUnlock(hCd)
    ctypes.windll.user32.SetClipboardData(1,hCd)
    ctypes.windll.user32.CloseClipboard()

def macSetClipboard(text):
    outf = os.popen('pbcopy', 'w')
    outf.write(smart_str(text))
    outf.close()

def macGetClipboard():
    outf = os.popen('pbpaste', 'r')
    content = outf.read()
    outf.close()
    return content

def gtkGetClipboard():
    return gtk.Clipboard().wait_for_text()

def gtkSetClipboard(text):
    cb = gtk.Clipboard()
    cb.set_text(smart_str(text))
    cb.store()

def qtGetClipboard():
    return str(cb.text())

def qtSetClipboard(text):
    cb.setText(smart_str(text))

def xclipSetClipboard(text):
    outf = os.popen('xclip -selection c', 'w')
    outf.write(smart_str(text))
    outf.close()

def xclipGetClipboard():
    outf = os.popen('xclip -selection c -o', 'r')
    content = outf.read()
    outf.close()
    return content

def xselSetClipboard(text):
    outf = os.popen('xsel -i', 'w')
    outf.write(smart_str(text))
    outf.close()

def xselGetClipboard():
    outf = os.popen('xsel -o', 'r')
    content = outf.read()
    outf.close()
    return content

if os.name == 'nt' or platform.system() == 'Windows':
    import ctypes
    getcb = winGetClipboard
    setcb = winSetClipboard
elif os.name == 'mac' or platform.system() == 'Darwin':
    getcb = macGetClipboard
    setcb = macSetClipboard
elif os.name == 'posix' or platform.system() == 'Linux':
    xclipExists = os.system('which xclip') == 0
    if xclipExists:
        getcb = xclipGetClipboard
        setcb = xclipSetClipboard
    else:
        xselExists = os.system('which xsel') == 0
        if xselExists:
            getcb = xselGetClipboard
            setcb = xselSetClipboard
        try:
            import gtk
            getcb = gtkGetClipboard
            setcb = gtkSetClipboard
        except:
            try:
                import PyQt4.QtCore
                import PyQt4.QtGui
                app = QApplication([])
                cb = PyQt4.QtGui.QApplication.clipboard()
                getcb = qtGetClipboard
                setcb = qtSetClipboard
            except:
                raise Exception(
'Pyperclip requires the gtk or PyQt4 module installed, or the xclip command.'
                )

copy = setcb
paste = getcb


src/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# 1337 Translator - Translates English text to 1337.
# Copyright (C) 2012 Vinny Diehl
#
# This application is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This application is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this application.  If not, see <http://www.gnu.org/licenses/>.

# Constants:
_version = '0.2'
_channel = 'trunk'

from argparse import ArgumentParser
from colorama import Fore, init
from json import loads
from pyperclip import copy, paste
from random import choice
from re import findall, sub

### Initialization

# Initialise colorama.
init()
_l_color = Fore.RED
_w_color = Fore.GREEN
_s_color = Fore.MAGENTA
_reset_color = Fore.RESET

# Parse command line options...

opt = ArgumentParser(description='Translate English to 1337.')

opt.add_argument('--version', action='store_true',
                 help='Display the current version number and exit.')

# It can only take input from one place, so make these mutually exclusive.
# If no input is selected, it will default to taking input from the keyboard.
input_group = opt.add_argument_group('Input', '''Choose a method of
                                     input. If none are selected, it will
                                     prompt you at runtime.''')
input_options = input_group.add_mutually_exclusive_group()
input_options.add_argument('-i', '--input',
                           help='Specify a string to be translated.')
input_options.add_argument('-c', '--clipboard', action='store_true',
                            help='Take input from the clipboard.')
input_options.add_argument('-f', '--file',
                            help='Take input from a file.')

# Add options for fake letters and/or words
fake_options = opt.add_argument_group('Fakeouts', '''Make your text harder to
                                       be read by outsiders.''')
fake_options.add_argument('-l', '--fake-letters', type=int, metavar='INTERVAL',
                           help='Add in a phoney letter at an interval.')
fake_options.add_argument('-w', '--fake-words', type=int, metavar='INTERVAL',
                           help='Add in a phoney word at an interval.')
fake_options.add_argument('-s', '--full-stop', action='store_true',
                           help="Translate '.' to STOP, as in a telegraph.")

# It can output to more than one place, so allow multiple output options.
output_options = opt.add_argument_group('Output', '''Select where to output
                                         the translated text.''')
output_options.add_argument('-C', '--out-clipboard', action='store_true',
                             help='Set the translated text to the clipboard.')
output_options.add_argument('-o', '--output', metavar='FILE',
                             help='Select a file to output the translation to.')

# This is useful if they're parsing a big file, outputting the translation
# to another file, and don't want the whole thing thrown in their face.
opt.add_argument('-q', '--quiet', '--silent', action='store_true',
                  help='Suppress all text output to the console.')

opt.add_argument('-t', '--no-trim', '--no-strip', action='store_true',
                 help='Do not trim leading/trailing whitespace from input.')

# Parse all of the arguments to args.
args = opt.parse_args()

# Print the version and exit if they asked for it
if args.version:
    print '1337 Translator'
    print 'Version: ' + _version
    print 'Release Channel: ' + _channel
    exit(0)

# Read in alphabet from JSON file. The alphabet is a dictionary, with each
# (lower case) letter of the English alphabet being a key. The value of each
# key is a list containing all of the possible translation of that letter.
leetalpha = loads(open('alphabet.json').read())

### Input

if args.input:
    raw = args.input
elif args.clipboard:
    # Get a string from the user's clipboard using Pyperclip
    raw = paste()
elif args.file:
    raw = open(args.file).read()
else:
    raw = raw_input('Enter the string to be translated: ')
    print

# We don't want to do pointless translations.
if not raw.strip():
    print 'There is no text to translate in the input data.'
    exit(1)

### Preprocessor

# Get rid of leading/trailing whitespace.
if not args.no_trim:
    raw = raw.strip()

# Show the data that has been read in.
if not args.quiet:
    print 'Original input:\n\n' + raw + '\n'

if args.fake_words:
    # Read in our list of possible words
    words = open('fakewords.txt').read().split('\n')

    # :KLUDGE: 2012-05-31 gbchaosmaster - Regex doesn't capture the last word
    # I tried everything and can't get this to work. I thought about just
    # having a second regex capture the last word but that'd just be sloppy.
    # For now, extremely hacky solution to get the part that the regex missed.
    tokens = findall(r'.+?[\s\n]+(?=[\w(])', raw)
    tokens.append(raw[len(''.join(tokens)):])

    # Stuff for counting and appending:
    i, fake = 0, ''
    for token in tokens:
        if i % args.fake_words == 0 and i > 0:
            fake += _w_color + choice(words) + _reset_color + ' '
        fake += token
        i += 1

    # Update raw to reflect the changes
    raw = fake

    if not args.quiet:
        print 'String after fake word preprocessing:\n\n' + raw + '\n'

    # Clear colors for --fake-words
    raw = raw.replace(_w_color, '').replace(_reset_color, '')

if args.full_stop:
    fake = sub(r' ?\. ?',
               _s_color + r' STOP ' + _reset_color,
               raw)

    # Trim the last space if the last character in the string was a full stop.
    if fake.endswith(' STOP ') and not raw.endswith(' STOP '):
        fake = fake[:-1]

    # Only display this if something changed. Otherwise, don't bother.
    if fake != raw:
        raw = fake
        if not args.quiet:
            print 'String after full stop preprocessing:\n\n' + raw + '\n'

    # Clear colors for --full-stop
    raw = raw.replace(_s_color, '').replace(_reset_color, '')

if args.fake_letters:
    # Put in a random character from engalpha every interval. We'll have to
    # count them manually. Same deal as above:
    i, fake = 0, ''
    for ch in raw:
        if ch.isalpha():
            if i % args.fake_letters == 0 and i > 0:
                # Pop in a random letter
                fake += _l_color + \
                        choice('abcdefghijklmnopqrstuvwxyz') + \
                        _reset_color
            # Only increment if it's a letter, symbols aren't counted.
            i += 1
        fake += ch

    raw = fake

    if not args.quiet:
        print 'String after fake letters preprocessing:\n\n' + raw + '\n'

    # Clear colors for --fake-letters
    raw = raw.replace(_l_color, '').replace(_reset_color, '')

# Lower case everything to prepare for translation.
raw = raw.lower()

### Translator

# Translate the entire string into 1337 by iterating over each character
# and translating that character only if it is in the English alphabet.
output = ''.join(choice(leetalpha[ch]) if ch.isalpha() else ch for ch in raw)

### Output

if args.out_clipboard:
    # Set the output to the clipboard using Pyperclip.
    copy(output)
if args.output:
    # Write the output to the specified file.
    open(args.output, 'w').write(output.encode('utf-8'))

if not args.quiet:
    print 'Translated text:\n\n' + output + '\n'
